// @applitools/image/dist/formats/png-sharp.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toPngSharp = exports.fromPngBufferSharp = void 0;
const sharp_1 = __importDefault(require("sharp"));
const buffer_1 = require("buffer");
async function fromPngBufferSharp(buffer) {
    // Ensure we always get RGBA (4 channels) output for consistency
    const { data, info } = await (0, sharp_1.default)(buffer_1.Buffer.from(buffer))
        .ensureAlpha() // Force RGBA output
        .raw()
        .toBuffer({ resolveWithObject: true });
    return {
        width: info.width,
        height: info.height,
        data: data,
    };
}
exports.fromPngBufferSharp = fromPngBufferSharp;
async function toPngSharp(image) {
    // Calculate channels from data size (only support RGB=3 and RGBA=4)
    const calculatedChannels = image.data.length / (image.width * image.height);
    const channels = Math.round(calculatedChannels);
    // Validate channels and data size
    if (![3, 4].includes(channels) || Math.abs(channels - calculatedChannels) > 0.01) {
        throw new Error(`Invalid image data: expected ${image.width}x${image.height} with RGB (3) or RGBA (4) channels, ` +
            `got ${image.data.length} bytes (${calculatedChannels.toFixed(2)} channels)`);
    }
    return await (0, sharp_1.default)(buffer_1.Buffer.from(image.data), {
        raw: {
            width: image.width,
            height: image.height,
            channels,
        },
    })
        .png({
        compressionLevel: 9,
        adaptiveFiltering: false,
    })
        .toBuffer();
}
exports.toPngSharp = toPngSharp;
