// @applitools/eyes/dist/enums/IosDeviceName.js
"use strict";
/* GENERATED FILE */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IosDeviceNameEnum = void 0;
var IosDeviceNameEnum;
(function (IosDeviceNameEnum) {
    IosDeviceNameEnum["iPad_7"] = "iPad (7th generation)";
    IosDeviceNameEnum["iPad_9"] = "iPad (9th generation)";
    IosDeviceNameEnum["iPad_10"] = "iPad (10th generation)";
    IosDeviceNameEnum["iPad_mini_6"] = "iPad mini (6th generation)";
    IosDeviceNameEnum["iPad_Air_4"] = "iPad Air (4th generation)";
    IosDeviceNameEnum["iPad_Air_11_inch_M3"] = "iPad Air 11-inch (M3)";
    IosDeviceNameEnum["iPad_Air_13_inch_M3"] = "iPad Air 13-inch (M3)";
    IosDeviceNameEnum["iPad_A16"] = "iPad (A16)";
    /** @deprecated use {@link iPad_Pro_12_9_inch_3 } instead */
    IosDeviceNameEnum["iPad_Pro_3"] = "iPad Pro (12.9-inch) (3rd generation)";
    /** @deprecated use {@link iPad_Pro_11_inch_4 } instead */
    IosDeviceNameEnum["iPad_Pro_4"] = "iPad Pro (11-inch) (4th generation)";
    IosDeviceNameEnum["iPad_Pro_11_inch_4"] = "iPad Pro (11-inch) (4th generation)";
    IosDeviceNameEnum["iPad_Pro_11_inch_M5"] = "iPad Pro 11-inch (M5)";
    IosDeviceNameEnum["iPad_Pro_13_inch_M5"] = "iPad Pro 13-inch (M5)";
    IosDeviceNameEnum["iPad_Pro_12_9_inch_3"] = "iPad Pro (12.9-inch) (3rd generation)";
    IosDeviceNameEnum["iPhone_SE_2"] = "iPhone SE (2nd generation)";
    IosDeviceNameEnum["iPhone_SE_3"] = "iPhone SE (3rd generation)";
    IosDeviceNameEnum["iPhone_XR"] = "iPhone XR";
    /** @deprecated use {@link iPhone_Xs } instead */
    IosDeviceNameEnum["iPhone_XS"] = "iPhone Xs";
    IosDeviceNameEnum["iPhone_Xs"] = "iPhone Xs";
    IosDeviceNameEnum["iPhone_11"] = "iPhone 11";
    IosDeviceNameEnum["iPhone_11_Pro_Max"] = "iPhone 11 Pro Max";
    IosDeviceNameEnum["iPhone_11_Pro"] = "iPhone 11 Pro";
    IosDeviceNameEnum["iPhone_12"] = "iPhone 12";
    IosDeviceNameEnum["iPhone_12_mini"] = "iPhone 12 mini";
    IosDeviceNameEnum["iPhone_12_Pro_Max"] = "iPhone 12 Pro Max";
    IosDeviceNameEnum["iPhone_12_Pro"] = "iPhone 12 Pro";
    IosDeviceNameEnum["iPhone_13"] = "iPhone 13";
    IosDeviceNameEnum["iPhone_13_mini"] = "iPhone 13 mini";
    IosDeviceNameEnum["iPhone_13_Pro_Max"] = "iPhone 13 Pro Max";
    IosDeviceNameEnum["iPhone_13_Pro"] = "iPhone 13 Pro";
    IosDeviceNameEnum["iPhone_14"] = "iPhone 14";
    IosDeviceNameEnum["iPhone_14_Pro_Max"] = "iPhone 14 Pro Max";
    IosDeviceNameEnum["iPhone_14_Pro"] = "iPhone 14 Pro";
    IosDeviceNameEnum["iPhone_14_Plus"] = "iPhone 14 Plus";
    IosDeviceNameEnum["iPhone_15"] = "iPhone 15";
    IosDeviceNameEnum["iPhone_15_Pro_Max"] = "iPhone 15 Pro Max";
    IosDeviceNameEnum["iPhone_15_Pro"] = "iPhone 15 Pro";
    IosDeviceNameEnum["iPhone_15_Plus"] = "iPhone 15 Plus";
    IosDeviceNameEnum["iPhone_16"] = "iPhone 16";
    IosDeviceNameEnum["iPhone_16_Pro_Max"] = "iPhone 16 Pro Max";
    IosDeviceNameEnum["iPhone_16_Pro"] = "iPhone 16 Pro";
    IosDeviceNameEnum["iPhone_16_Plus"] = "iPhone 16 Plus";
    IosDeviceNameEnum["iPhone_17"] = "iPhone 17";
    IosDeviceNameEnum["iPhone_17_Pro"] = "iPhone 17 Pro";
    IosDeviceNameEnum["iPhone_17_Pro_Max"] = "iPhone 17 Pro Max";
    IosDeviceNameEnum["iPhone_Air"] = "iPhone Air";
})(IosDeviceNameEnum || (exports.IosDeviceNameEnum = IosDeviceNameEnum = {}));
