// @applitools/eyes/dist/enums/IosMultiDeviceTarget.js
"use strict";
/* GENERATED FILE */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IosMultiDeviceTarget = void 0;
const IosDeviceTarget_1 = require("../input/IosDeviceTarget");
// Const object with getters that return Device Target instances for fluent API
exports.IosMultiDeviceTarget = {
    get iPhone_SE_1() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone SE (1st generation)');
    },
    get iPhone_SE_2() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone SE (2nd generation)');
    },
    get iPhone_SE_3() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone SE (3rd generation)');
    },
    get iPhone_6() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 6');
    },
    get iPhone_6_Plus() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 6 Plus');
    },
    get iPhone_7() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 7');
    },
    get iPhone_7_Plus() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 7 Plus');
    },
    get iPhone_8() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 8');
    },
    get iPhone_8_Plus() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 8 Plus');
    },
    get iPhone_X() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone X');
    },
    get iPhone_XR() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone XR');
    },
    get iPhone_Xs() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone Xs');
    },
    get iPhone_Xs_Max() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone Xs Max');
    },
    get iPhone_11() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 11');
    },
    get iPhone_11_Pro() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 11 Pro');
    },
    get iPhone_11_Pro_Max() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 11 Pro Max');
    },
    get iPhone_12() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 12');
    },
    get iPhone_12_mini() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 12 mini');
    },
    get iPhone_12_Pro() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 12 Pro');
    },
    get iPhone_12_Pro_Max() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 12 Pro Max');
    },
    get iPhone_13() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 13');
    },
    get iPhone_13_mini() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 13 mini');
    },
    get iPhone_13_Pro() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 13 Pro');
    },
    get iPhone_13_Pro_Max() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 13 Pro Max');
    },
    get iPhone_14() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 14');
    },
    get iPhone_14_Pro() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 14 Pro');
    },
    get iPhone_14_Plus() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 14 Plus');
    },
    get iPhone_14_Pro_Max() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 14 Pro Max');
    },
    get iPhone_15() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 15');
    },
    get iPhone_15_Pro() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 15 Pro');
    },
    get iPhone_15_Plus() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 15 Plus');
    },
    get iPhone_15_Pro_Max() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 15 Pro Max');
    },
    get iPhone_16() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 16');
    },
    get iPhone_16e() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 16e');
    },
    get iPhone_16_Pro() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 16 Pro');
    },
    get iPhone_16_Plus() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 16 Plus');
    },
    get iPhone_16_Pro_Max() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 16 Pro Max');
    },
    get iPhone_17() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 17');
    },
    get iPhone_17_Pro() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 17 Pro');
    },
    get iPhone_17_Pro_Max() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone 17 Pro Max');
    },
    get iPhone_Air() {
        return new IosDeviceTarget_1.IosDeviceTarget('iPhone Air');
    },
};
