// @applitools/dom-shared/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAccessibleFrame = exports.isInlineFrame = exports.absolutizeUrl = exports.poll = exports.pollify = exports.chunkify = void 0;
const chunkify_1 = __importDefault(require("./chunkify"));
exports.chunkify = chunkify_1.default;
const pollify_1 = __importDefault(require("./pollify"));
exports.pollify = pollify_1.default;
const poll_1 = __importDefault(require("./poll"));
exports.poll = poll_1.default;
const absolutizeUrl_1 = __importDefault(require("./browser/absolutizeUrl"));
exports.absolutizeUrl = absolutizeUrl_1.default;
const isInlineFrame_1 = __importDefault(require("./browser/isInlineFrame"));
exports.isInlineFrame = isInlineFrame_1.default;
const isAccessibleFrame_1 = __importDefault(require("./browser/isAccessibleFrame"));
exports.isAccessibleFrame = isAccessibleFrame_1.default;
