// @applitools/core/dist/ufg/utils/take-dom-snapshot.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.takeDomSnapshot = void 0;
const utils = __importStar(require("@applitools/utils"));
const { processPagePoll, pollResult } = require('@applitools/dom-snapshot');
async function takeDomSnapshot({ context, settings, logger, }) {
    var _a, _b;
    const driver = context.driver;
    const environment = await driver.getEnvironment();
    const features = await driver.getFeatures();
    const cookies = (_a = settings === null || settings === void 0 ? void 0 : settings.cookies) !== null && _a !== void 0 ? _a : [];
    // assumption is that elementReferences are all in the frame which is `context`
    // otherwise, the elements would still be found inside other frames, but `takeContextDomSnapshot` should be passed only the specific element references that it needs
    const elements = [];
    for (const elementReference of (_b = settings === null || settings === void 0 ? void 0 : settings.elementReferences) !== null && _b !== void 0 ? _b : []) {
        elements.push(await context.elements(elementReference));
    }
    const snapshot = transformRawDomSnapshot(await takeContextDomSnapshot({ context, isRootFrame: true }), elements);
    snapshot.cookies = features.allCookies ? await driver.getCookies().catch(() => []) : cookies; // intentionally fetching cookies after the dom snapshot runs
    return snapshot;
    async function takeContextDomSnapshot({ context, isRootFrame, }) {
        // logger.log(`taking dom snapshot. ${context._reference ? `context referece: ${JSON.stringify(context._reference)}` : ''}`)
        var _a, _b, _c, _d;
        const arg = {
            dontFetchResources: settings === null || settings === void 0 ? void 0 : settings.disableBrowserFetching,
            skipResources: settings === null || settings === void 0 ? void 0 : settings.skipResources,
            fetchConcurrency: settings === null || settings === void 0 ? void 0 : settings.fetchConcurrency,
            removeReverseProxyURLPrefixes: Boolean(process.env.APPLITOOLS_SCRIPT_REMOVE_REVERSE_PROXY_URL_PREFIXES),
            chunkByteLength: (_a = settings === null || settings === void 0 ? void 0 : settings.chunkByteLength) !== null && _a !== void 0 ? _a : (Number(process.env.APPLITOOLS_SCRIPT_RESULT_MAX_BYTE_LENGTH) ||
                (environment.isIOS ? 100000 : 250 * 1024 * 1024)),
            serializeResources: true,
            compressResources: false,
            showLogs: settings === null || settings === void 0 ? void 0 : settings.showLogs,
            elements: isRootFrame ? elements : [],
        };
        const scripts = {
            main: features.canExecuteOnlyFunctionScripts
                ? processPagePoll
                : `/* @applitools/dom-snapshot@1 */ return (${processPagePoll}).apply(null, arguments);`,
            poll: features.canExecuteOnlyFunctionScripts
                ? pollResult
                : `/* @applitools/dom-snapshot@1 */ return (${pollResult}).apply(null, arguments);`,
        };
        const snapshot = await context.executePoll(scripts, {
            main: arg,
            poll: arg,
            executionTimeout: (_b = settings === null || settings === void 0 ? void 0 : settings.executionTimeout) !== null && _b !== void 0 ? _b : 5 * 60 * 1000,
            pollTimeout: (_c = settings === null || settings === void 0 ? void 0 : settings.pollTimeout) !== null && _c !== void 0 ? _c : 200,
        });
        // intentionally fetching cookies after the dom snapshot runs
        if (!features.allCookies) {
            cookies.push(...(await context.getCookies()));
        }
        const crossFrames = extractCrossFrames({ snapshot, logger });
        for (const { reference, parentSnapshot, cdtNode } of crossFrames) {
            const frameContext = await context
                .context(reference)
                .then(context => context.focus())
                .catch(err => {
                var _a;
                const srcAttr = (_a = cdtNode.attributes) === null || _a === void 0 ? void 0 : _a.find(attr => attr.name === 'src');
                if (srcAttr)
                    srcAttr.value = '';
                logger.log(`could not switch to frame during takeDomSnapshot. Path to frame: ${JSON.stringify(reference)}`, err);
            });
            if (frameContext) {
                const frameSnapshot = await takeContextDomSnapshot({
                    context: frameContext,
                    isRootFrame: false, // when we're diving into a frame, it's not the root frame anymore and we shouldn't pass the elements
                });
                let url = new URL(frameSnapshot.url);
                if (url.protocol === 'data:')
                    url = new URL(`http://data-url-frame${url.search}`);
                if (!url.searchParams.has('applitools-iframe'))
                    url.searchParams.set('applitools-iframe', utils.general.guid());
                frameSnapshot.url = url.href;
                parentSnapshot.frames.push(frameSnapshot);
                (_d = cdtNode.attributes) === null || _d === void 0 ? void 0 : _d.push({ name: 'data-applitools-src', value: frameSnapshot.url });
            }
        }
        logSnapshot(logger, snapshot);
        return snapshot;
    }
}
exports.takeDomSnapshot = takeDomSnapshot;
function transformRawDomSnapshot(snapshot, elements) {
    const { blobs, safeSelectors, selector: _, crossFrames: __, ...rest } = snapshot;
    const domSnapshot = {
        ...rest,
        resourceContents: blobs.reduce((resourceContents, blob) => {
            return {
                ...resourceContents,
                [blob.url]: blob,
            };
        }, {}),
        frames: snapshot.frames.map(frameSnapshot => transformRawDomSnapshot(frameSnapshot)),
    };
    if (elements === null || elements === void 0 ? void 0 : elements.length) {
        domSnapshot.generatedSelectors = elements.map((elements, i) => {
            var _a, _b;
            return ({
                originalSelector: (_b = (_a = elements[0]) === null || _a === void 0 ? void 0 : _a.commonSelector) !== null && _b !== void 0 ? _b : null,
                safeSelector: safeSelectors[i],
            });
        });
    }
    return domSnapshot;
}
function extractCrossFrames({ snapshot, parent = null, logger, }) {
    var _a;
    const crossFrames = [snapshot, ...((_a = snapshot.frames) !== null && _a !== void 0 ? _a : [])].flatMap((snapshot, index) => {
        var _a;
        const crossFrames = ((_a = snapshot.crossFrames) !== null && _a !== void 0 ? _a : []).map(({ selector, index }) => ({
            reference: { reference: { type: 'css', selector }, parent },
            parentSnapshot: snapshot,
            cdtNode: snapshot.cdt[index],
        }));
        return [
            ...crossFrames,
            ...(index > 0
                ? extractCrossFrames({
                    snapshot,
                    parent: { reference: { type: 'css', selector: snapshot.selector }, parent },
                    logger,
                })
                : []),
        ];
    });
    logger.log(...(snapshot.crossFrames
        ? [
            `dom snapshot - frames paths for ${JSON.stringify(snapshot.crossFrames)}:`,
            crossFrames.map(selector => ({
                // we don't want `selector.parentSnapshot` to be logged - the size of it can be huge
                cdtNode: selector.cdtNode,
                reference: selector.reference,
            })),
        ]
        : ['dom snapshot - no cross frames found']));
    return crossFrames;
}
function logSnapshot(logger, snapshot) {
    try {
        logger.log(`dom snapshot - cdt length: ${snapshot.cdt.length}`);
        logger.log(`dom snapshot - blobs urls (${snapshot.blobs.length}):`, JSON.stringify(snapshot.blobs.map(({ url }) => url)));
        logger.log(`dom snapshot - resource urls (${snapshot.resourceUrls.length}):`, JSON.stringify(snapshot.resourceUrls));
        logger.log(`dom snapshot - cdt nodes with data-applitools-selector: ${JSON.stringify(snapshot.cdt.filter(node => { var _a; return (_a = node.attributes) === null || _a === void 0 ? void 0 : _a.some(attr => attr.name === 'data-applitools-selector'); }))}`);
        logger.log('dom snapshot - safeSelectors:', snapshot.safeSelectors);
        logger.log('dom snapshot - warnings:', snapshot.warnings);
        snapshot.errors && logger.log('dom snapshot - errors:', snapshot.errors);
    }
    catch (err) {
        logger.error(`error while logging dom snapshot`, err);
    }
}
