// @applitools/ufg-client/dist/offline/chrome-emulation-devices.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chromeEmulationDevices = void 0;
exports.chromeEmulationDevices = {
    'iPhone 4': {
        portrait: {
            height: 480,
            width: 320,
        },
        landscape: {
            height: 320,
            width: 480,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPhone 5/SE': {
        portrait: {
            height: 568,
            width: 320,
        },
        landscape: {
            height: 320,
            width: 568,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPhone 6/7/8': {
        portrait: {
            height: 667,
            width: 375,
        },
        landscape: {
            height: 375,
            width: 667,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPhone 6/7/8 Plus': {
        portrait: {
            height: 736,
            width: 414,
        },
        landscape: {
            height: 414,
            width: 736,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPhone X': {
        portrait: {
            height: 812,
            width: 375,
        },
        landscape: {
            height: 375,
            width: 812,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'BlackBerry Z30': {
        portrait: {
            height: 640,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 640,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Nexus 4': {
        portrait: {
            height: 640,
            width: 384,
        },
        landscape: {
            height: 384,
            width: 640,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Nexus 5': {
        portrait: {
            height: 640,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 640,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Nexus 5X': {
        portrait: {
            height: 732,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 732,
        },
        metadata: {
            pixelRatio: 2.625,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Nexus 6': {
        portrait: {
            height: 732,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 732,
        },
        metadata: {
            pixelRatio: 3.5,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Nexus 6P': {
        portrait: {
            height: 732,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 732,
        },
        metadata: {
            pixelRatio: 3.5,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Pixel 2': {
        portrait: {
            height: 731,
            width: 411,
        },
        landscape: {
            height: 411,
            width: 731,
        },
        metadata: {
            pixelRatio: 2.625,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Pixel 2 XL': {
        portrait: {
            height: 823,
            width: 411,
        },
        landscape: {
            height: 411,
            width: 823,
        },
        metadata: {
            pixelRatio: 3.5,
            capabilities: ['touch', 'mobile'],
        },
    },
    'LG Optimus L70': {
        portrait: {
            height: 640,
            width: 384,
        },
        landscape: {
            height: 384,
            width: 640,
        },
        metadata: {
            pixelRatio: 1.25,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Nokia N9': {
        portrait: {
            height: 854,
            width: 480,
        },
        landscape: {
            height: 480,
            width: 854,
        },
        metadata: {
            pixelRatio: 1,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Nokia Lumia 520': {
        portrait: {
            height: 533,
            width: 320,
        },
        landscape: {
            height: 320,
            width: 533,
        },
        metadata: {
            pixelRatio: 1.5,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Microsoft Lumia 550': {
        portrait: {
            height: 360,
            width: 640,
        },
        landscape: {
            height: 360,
            width: 640,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Microsoft Lumia 950': {
        portrait: {
            height: 640,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 640,
        },
        metadata: {
            pixelRatio: 4,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S3': {
        portrait: {
            height: 640,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 640,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S III': {
        portrait: {
            height: 640,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 640,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S5': {
        portrait: {
            height: 640,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 640,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Kindle Fire HDX': {
        portrait: {
            height: 1280,
            width: 800,
        },
        landscape: {
            height: 800,
            width: 1280,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPad Mini': {
        portrait: {
            height: 1024,
            width: 768,
        },
        landscape: {
            height: 768,
            width: 1024,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    iPad: {
        portrait: {
            height: 1024,
            width: 768,
        },
        landscape: {
            height: 768,
            width: 1024,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPad Pro': {
        portrait: {
            height: 1366,
            width: 1024,
        },
        landscape: {
            height: 1024,
            width: 1366,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Blackberry PlayBook': {
        portrait: {
            height: 1024,
            width: 600,
        },
        landscape: {
            height: 600,
            width: 1024,
        },
        metadata: {
            pixelRatio: 1,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Nexus 10': {
        portrait: {
            height: 1296,
            width: 968,
        },
        landscape: {
            height: 968,
            width: 1296,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Nexus 7': {
        portrait: {
            height: 960,
            width: 600,
        },
        landscape: {
            height: 600,
            width: 960,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy Note 3': {
        portrait: {
            height: 640,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 640,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy Note II': {
        portrait: {
            height: 640,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 640,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy Note 2': {
        portrait: {
            height: 640,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 640,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S20': {
        portrait: {
            height: 800,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 800,
        },
        metadata: {
            pixelRatio: 4,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S23': {
        portrait: {
            height: 780,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 780,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S22': {
        portrait: {
            height: 780,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 780,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S21': {
        portrait: {
            height: 800,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 800,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S21 Ultra': {
        portrait: {
            height: 800,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 800,
        },
        metadata: {
            pixelRatio: 4,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S22 Ultra': {
        portrait: {
            height: 772,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 772,
        },
        metadata: {
            pixelRatio: 4,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S23 Ultra': {
        portrait: {
            height: 824,
            width: 384,
        },
        landscape: {
            height: 384,
            width: 824,
        },
        metadata: {
            pixelRatio: 3.75,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S25': {
        portrait: {
            height: 780,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 780,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S25 Ultra': {
        portrait: {
            height: 824,
            width: 384,
        },
        landscape: {
            height: 384,
            width: 824,
        },
        metadata: {
            pixelRatio: 3.75,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Laptop with touch': {
        portrait: {
            height: 1280,
            width: 950,
        },
        landscape: {
            height: 950,
            width: 1280,
        },
        metadata: {
            pixelRatio: 1,
            capabilities: ['touch'],
        },
    },
    'Laptop with HiDPI screen': {
        portrait: {
            height: 1440,
            width: 900,
        },
        landscape: {
            height: 900,
            width: 1440,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: [],
        },
    },
    'Laptop with MDPI screen': {
        portrait: {
            height: 1280,
            width: 800,
        },
        landscape: {
            height: 800,
            width: 1280,
        },
        metadata: {
            pixelRatio: 1,
            capabilities: [],
        },
    },
    'iPhone XR': {
        portrait: {
            height: 808,
            width: 414,
        },
        landscape: {
            height: 414,
            width: 808,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPhone XS Max': {
        portrait: {
            height: 808,
            width: 414,
        },
        landscape: {
            height: 414,
            width: 808,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPhone XS': {
        portrait: {
            height: 724,
            width: 375,
        },
        landscape: {
            height: 375,
            width: 724,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy A52s': {
        portrait: {
            height: 915,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 915,
        },
        metadata: {
            pixelRatio: 2.62,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Samsung Galaxy A5': {
        portrait: {
            height: 640,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 640,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy A5': {
        portrait: {
            height: 640,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 640,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Samsung Galaxy S8': {
        portrait: {
            height: 692,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 692,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S8': {
        portrait: {
            height: 692,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 692,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'LG G6': {
        portrait: {
            height: 720,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 720,
        },
        metadata: {
            pixelRatio: 4,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPad Air 2': {
        portrait: {
            height: 1024,
            width: 768,
        },
        landscape: {
            height: 768,
            width: 1024,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPad 6th Gen': {
        portrait: {
            height: 1024,
            width: 768,
        },
        landscape: {
            height: 768,
            width: 1024,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPhone 14 Pro': {
        portrait: {
            height: 852,
            width: 393,
        },
        landscape: {
            height: 393,
            width: 852,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPhone SE 3rd Gen': {
        portrait: {
            height: 667,
            width: 375,
        },
        landscape: {
            height: 375,
            width: 667,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPhone 14 Plus': {
        portrait: {
            height: 926,
            width: 428,
        },
        landscape: {
            height: 428,
            width: 926,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPhone 15 Pro': {
        portrait: {
            height: 852,
            width: 393,
        },
        landscape: {
            height: 393,
            width: 852,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPhone 13 Mini': {
        portrait: {
            height: 812,
            width: 375,
        },
        landscape: {
            height: 375,
            width: 812,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPhone 11': {
        portrait: {
            height: 808,
            width: 414,
        },
        landscape: {
            height: 414,
            width: 808,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPhone 11 Pro Max': {
        portrait: {
            height: 808,
            width: 414,
        },
        landscape: {
            height: 414,
            width: 808,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPhone 11 Pro': {
        portrait: {
            height: 724,
            width: 375,
        },
        landscape: {
            height: 375,
            width: 724,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S10': {
        portrait: {
            height: 778,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 778,
        },
        metadata: {
            pixelRatio: 2.625,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S9 Plus': {
        portrait: {
            height: 798,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 798,
        },
        metadata: {
            pixelRatio: 2.625,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S9': {
        portrait: {
            height: 692,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 692,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S10 Plus': {
        portrait: {
            height: 780,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 780,
        },
        metadata: {
            pixelRatio: 2.625,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy S8 Plus': {
        portrait: {
            height: 798,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 798,
        },
        metadata: {
            pixelRatio: 2.625,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy Note 10': {
        portrait: {
            height: 787,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 787,
        },
        metadata: {
            pixelRatio: 2.625,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy Note 10 Plus': {
        portrait: {
            height: 789,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 789,
        },
        metadata: {
            pixelRatio: 2.625,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy Note 9': {
        portrait: {
            height: 798,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 798,
        },
        metadata: {
            pixelRatio: 2.625,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy Note 8': {
        portrait: {
            height: 798,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 798,
        },
        metadata: {
            pixelRatio: 2.625,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy Note 4': {
        portrait: {
            height: 640,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 640,
        },
        metadata: {
            pixelRatio: 4,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Pixel 3': {
        portrait: {
            height: 738,
            width: 393,
        },
        landscape: {
            height: 393,
            width: 738,
        },
        metadata: {
            pixelRatio: 2.75,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Pixel 3 XL': {
        portrait: {
            height: 749,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 749,
        },
        metadata: {
            pixelRatio: 3.5,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Pixel 4': {
        portrait: {
            height: 830,
            width: 393,
        },
        landscape: {
            height: 393,
            width: 830,
        },
        metadata: {
            pixelRatio: 2.75,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Pixel 4 XL': {
        portrait: {
            height: 869,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 869,
        },
        metadata: {
            pixelRatio: 3.5,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Pixel 5': {
        portrait: {
            height: 851,
            width: 393,
        },
        landscape: {
            height: 393,
            width: 851,
        },
        metadata: {
            pixelRatio: 2.75,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Pixel 9': {
        portrait: {
            height: 840,
            width: 390,
        },
        landscape: {
            height: 390,
            width: 840,
        },
        metadata: {
            pixelRatio: 2.75,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPad 7th Gen': {
        portrait: {
            height: 1080,
            width: 810,
        },
        landscape: {
            height: 810,
            width: 1080,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPad 8th Gen': {
        portrait: {
            height: 1080,
            width: 810,
        },
        landscape: {
            height: 810,
            width: 1080,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPad 10th Gen': {
        portrait: {
            height: 1180,
            width: 820,
        },
        landscape: {
            height: 820,
            width: 1180,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPad Mini 4th Gen': {
        portrait: {
            height: 1024,
            width: 768,
        },
        landscape: {
            height: 768,
            width: 1024,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'iPad Mini 6th Gen': {
        portrait: {
            height: 1133,
            width: 744,
        },
        landscape: {
            height: 744,
            width: 1133,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'OnePlus 7T': {
        portrait: {
            height: 841,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 841,
        },
        metadata: {
            pixelRatio: 2.625,
            capabilities: ['touch', 'mobile'],
        },
    },
    'OnePlus 7T Pro': {
        portrait: {
            height: 892,
            width: 412,
        },
        landscape: {
            height: 412,
            width: 892,
        },
        metadata: {
            pixelRatio: 3.5,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy Tab S7': {
        portrait: {
            height: 1280,
            width: 800,
        },
        landscape: {
            height: 800,
            width: 1280,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy Tab S8': {
        portrait: {
            height: 1280,
            width: 800,
        },
        landscape: {
            height: 800,
            width: 1280,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy Tab A8': {
        portrait: {
            height: 1280,
            width: 800,
        },
        landscape: {
            height: 800,
            width: 1280,
        },
        metadata: {
            pixelRatio: 1.5,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy Tab A SM-T510': {
        portrait: {
            height: 1280,
            width: 800,
        },
        landscape: {
            height: 800,
            width: 1280,
        },
        metadata: {
            pixelRatio: 1.5,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Galaxy Tab A7 SM-T500': {
        portrait: {
            height: 1335,
            width: 800,
        },
        landscape: {
            height: 800,
            width: 1335,
        },
        metadata: {
            pixelRatio: 1.5,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Sony Xperia 10 II': {
        portrait: {
            height: 840,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 840,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Huawei Mate 50 Pro': {
        portrait: {
            height: 980,
            width: 450,
        },
        landscape: {
            height: 450,
            width: 980,
        },
        metadata: {
            pixelRatio: 2.675,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Huawei Matepad 11': {
        portrait: {
            height: 930,
            width: 1485,
        },
        landscape: {
            height: 1485,
            width: 930,
        },
        metadata: {
            pixelRatio: 1.725,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Huawei P30': {
        portrait: {
            height: 720,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 720,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Huawei P40': {
        portrait: {
            height: 720,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 720,
        },
        metadata: {
            pixelRatio: 3,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Xiaomi Mi A3': {
        portrait: {
            height: 780,
            width: 360,
        },
        landscape: {
            height: 360,
            width: 780,
        },
        metadata: {
            pixelRatio: 2,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Xiaomi Redmi Note 8': {
        portrait: {
            height: 851,
            width: 393,
        },
        landscape: {
            height: 393,
            width: 851,
        },
        metadata: {
            pixelRatio: 2.75,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Xiaomi Mi Poco X3 Pro': {
        portrait: {
            height: 960,
            width: 432,
        },
        landscape: {
            height: 432,
            width: 960,
        },
        metadata: {
            pixelRatio: 2.5,
            capabilities: ['touch', 'mobile'],
        },
    },
    'Xiaomi Poco X3': {
        portrait: {
            height: 960,
            width: 432,
        },
        landscape: {
            height: 432,
            width: 960,
        },
        metadata: {
            pixelRatio: 2.5,
            capabilities: ['touch', 'mobile'],
        },
    },
};
