// @applitools/ufg-client/dist/offline/ios-devices.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.iosDevices = void 0;
exports.iosDevices = {
    'iPad Pro (11-inch) (4th generation)': {
        portrait: {
            height: 1194,
            width: 844,
        },
        landscapeLeft: {
            height: 844,
            width: 1194,
        },
        landscapeRight: {
            height: 844,
            width: 1194,
        },
        landscape: {
            height: 844,
            width: 1194,
        },
        metadata: {
            supportedIosVersions: {
                canary: '17',
                latest: '17',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 2,
        },
    },
    'iPad (9th generation)': {
        portrait: {
            height: 1080,
            width: 810,
        },
        landscapeLeft: {
            height: 810,
            width: 1080,
        },
        landscapeRight: {
            height: 810,
            width: 1080,
        },
        landscape: {
            height: 810,
            width: 1080,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                canary: '17',
                'canary-1': '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 2,
        },
    },
    'iPhone 15 Pro Max': {
        portrait: {
            height: 932,
            width: 430,
        },
        landscapeLeft: {
            height: 430,
            width: 932,
        },
        landscapeRight: {
            height: 430,
            width: 932,
        },
        landscape: {
            height: 430,
            width: 932,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                canary: '17',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone 15': {
        portrait: {
            height: 852,
            width: 393,
        },
        landscapeLeft: {
            height: 393,
            width: 852,
        },
        landscapeRight: {
            height: 393,
            width: 852,
        },
        landscape: {
            height: 393,
            width: 852,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                canary: '17',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone 14 Pro Max': {
        portrait: {
            height: 932,
            width: 430,
        },
        landscapeLeft: {
            height: 430,
            width: 932,
        },
        landscapeRight: {
            height: 430,
            width: 932,
        },
        landscape: {
            height: 430,
            width: 932,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                canary: '17',
                'canary-1': '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone 14': {
        portrait: {
            height: 844,
            width: 390,
        },
        landscapeLeft: {
            height: 390,
            width: 844,
        },
        landscapeRight: {
            height: 390,
            width: 844,
        },
        landscape: {
            height: 390,
            width: 844,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                canary: '17',
                'canary-1': '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone 13 Pro': {
        portrait: {
            height: 844,
            width: 390,
        },
        landscapeLeft: {
            height: 390,
            width: 844,
        },
        landscapeRight: {
            height: 390,
            width: 844,
        },
        landscape: {
            height: 390,
            width: 844,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                canary: '17',
                'canary-1': '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone 13': {
        portrait: {
            height: 844,
            width: 390,
        },
        landscapeLeft: {
            height: 390,
            width: 844,
        },
        landscapeRight: {
            height: 390,
            width: 844,
        },
        landscape: {
            height: 390,
            width: 844,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                canary: '17',
                'canary-1': '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone 13 Pro Max': {
        portrait: {
            height: 926,
            width: 428,
        },
        landscapeLeft: {
            height: 428,
            width: 926,
        },
        landscapeRight: {
            height: 428,
            width: 926,
        },
        landscape: {
            height: 428,
            width: 926,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                canary: '17',
                'canary-1': '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone 12 Pro Max': {
        portrait: {
            height: 926,
            width: 428,
        },
        landscapeLeft: {
            height: 428,
            width: 926,
        },
        landscapeRight: {
            height: 428,
            width: 926,
        },
        landscape: {
            height: 428,
            width: 926,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                canary: '17',
                'canary-1': '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone 12 Pro': {
        portrait: {
            height: 844,
            width: 390,
        },
        landscapeLeft: {
            height: 390,
            width: 844,
        },
        landscapeRight: {
            height: 390,
            width: 844,
        },
        landscape: {
            height: 390,
            width: 844,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                canary: '17',
                'canary-1': '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone 12': {
        portrait: {
            height: 844,
            width: 390,
        },
        landscapeLeft: {
            height: 390,
            width: 844,
        },
        landscapeRight: {
            height: 390,
            width: 844,
        },
        landscape: {
            height: 390,
            width: 844,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                canary: '17',
                'canary-1': '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone 12 mini': {
        portrait: {
            height: 780,
            width: 360,
        },
        landscapeLeft: {
            height: 360,
            width: 780,
        },
        landscapeRight: {
            height: 360,
            width: 780,
        },
        landscape: {
            height: 360,
            width: 780,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                canary: '17',
                'canary-1': '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone 11 Pro Max': {
        portrait: {
            height: 896,
            width: 414,
        },
        landscapeLeft: {
            height: 414,
            width: 896,
        },
        landscapeRight: {
            height: 414,
            width: 896,
        },
        landscape: {
            height: 414,
            width: 896,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                canary: '17',
                'canary-1': '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone 11 Pro': {
        portrait: {
            height: 812,
            width: 375,
        },
        landscapeLeft: {
            height: 375,
            width: 812,
        },
        landscapeRight: {
            height: 375,
            width: 812,
        },
        landscape: {
            height: 375,
            width: 812,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                canary: '17',
                'canary-1': '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone 11': {
        portrait: {
            height: 896,
            width: 414,
        },
        landscapeLeft: {
            height: 414,
            width: 896,
        },
        landscapeRight: {
            height: 414,
            width: 896,
        },
        landscape: {
            height: 414,
            width: 896,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                canary: '17',
                'canary-1': '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 2,
        },
    },
    'iPhone Xs': {
        portrait: {
            height: 812,
            width: 375,
        },
        landscapeLeft: {
            height: 375,
            width: 812,
        },
        landscapeRight: {
            height: 375,
            width: 812,
        },
        landscape: {
            height: 375,
            width: 812,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                'canary-1': '16',
                canary: '17',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone X': {
        portrait: {
            height: 812,
            width: 375,
        },
        landscapeLeft: {
            height: 375,
            width: 812,
        },
        landscapeRight: {
            height: 375,
            width: 812,
        },
        landscape: {
            height: 375,
            width: 812,
        },
        metadata: {
            supportedIosVersions: {
                latest: '16',
                canary: '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
    'iPhone XR': {
        portrait: {
            height: 896,
            width: 414,
        },
        landscapeLeft: {
            height: 414,
            width: 896,
        },
        landscapeRight: {
            height: 414,
            width: 896,
        },
        landscape: {
            height: 414,
            width: 896,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                'canary-1': '16',
                canary: '17',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 2,
        },
    },
    'iPhone 8': {
        portrait: {
            height: 667,
            width: 375,
        },
        landscapeLeft: {
            height: 375,
            width: 667,
        },
        landscapeRight: {
            height: 375,
            width: 667,
        },
        landscape: {
            height: 375,
            width: 667,
        },
        metadata: {
            supportedIosVersions: {
                latest: '16',
                canary: '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 2,
        },
    },
    'iPad Air (4th generation)': {
        portrait: {
            height: 1180,
            width: 820,
        },
        landscapeLeft: {
            height: 820,
            width: 1180,
        },
        landscapeRight: {
            height: 820,
            width: 1180,
        },
        landscape: {
            height: 820,
            width: 1180,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                'canary-1': '16',
                canary: '17',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 2,
        },
    },
    'iPad Pro (12.9-inch) (3rd generation)': {
        portrait: {
            height: 1366,
            width: 1024,
        },
        landscapeLeft: {
            height: 1024,
            width: 1366,
        },
        landscapeRight: {
            height: 1024,
            width: 1366,
        },
        landscape: {
            height: 1024,
            width: 1366,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                'canary-1': '16',
                canary: '17',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 2,
        },
    },
    'iPad (7th generation)': {
        portrait: {
            height: 1080,
            width: 810,
        },
        landscapeLeft: {
            height: 810,
            width: 1080,
        },
        landscapeRight: {
            height: 810,
            width: 1080,
        },
        landscape: {
            height: 810,
            width: 1080,
        },
        metadata: {
            supportedIosVersions: {
                latest: '17',
                'latest-1': '16',
                'canary-1': '16',
                canary: '17',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 2,
        },
    },
    'iPhone 8 Plus': {
        portrait: {
            height: 360,
            width: 640,
        },
        landscapeLeft: {
            height: 640,
            width: 360,
        },
        landscapeRight: {
            height: 640,
            width: 360,
        },
        landscape: {
            height: 640,
            width: 360,
        },
        metadata: {
            supportedIosVersions: {
                latest: '16',
                canary: '16',
            },
            supportedPlatforms: ['web', 'native'],
            pixelRatio: 3,
        },
    },
};
