// @applitools/core/dist/open-eyes.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleConcurrency = exports.getLatestCommitInfoFromEnvVars = exports.makeOpenEyes = void 0;
const driver_1 = require("@applitools/driver");
const core_1 = require("./classic/core");
const core_2 = require("./ufg/core");
const get_typed_eyes_1 = require("./get-typed-eyes");
const check_1 = require("./check");
const check_and_close_1 = require("./check-and-close");
const close_1 = require("./close");
const abort_1 = require("./abort");
const get_eyes_results_1 = require("./get-eyes-results");
const populate_eyes_server_settings_1 = require("./utils/populate-eyes-server-settings");
const utils = __importStar(require("@applitools/utils"));
const core_base_1 = require("@applitools/core-base");
const extract_git_info_1 = require("./utils/extract-git-info");
const types_1 = require("./types");
function makeOpenEyes({ type: defaultType = 'classic', clients, batch, removeDuplicateTests, core, cores, spec, environment, logger: mainLogger, asyncCache, concurrency, useServerConcurrency = false, cwd = process.cwd(), }) {
    let shouldPrintConcurrencyWarning = true;
    return async function openEyes({ type = defaultType, settings: openSettings, config, target, logger = mainLogger, }) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0;
        var _1, _2, _3, _4, _5;
        logger = logger.extend(mainLogger, { tags: [`eyes-${type}-${utils.general.shortid()}`] });
        logger.log(`Opening Eyes in directory "${cwd}"`);
        const settings = { environments: config === null || config === void 0 ? void 0 : config.check.environments, ...config === null || config === void 0 ? void 0 : config.open, ...openSettings };
        const eyesServerSettings = (0, populate_eyes_server_settings_1.populateEyesServerSettings)(settings);
        logger.mask(eyesServerSettings.apiKey);
        logger.mask(eyesServerSettings.eyesServerUrl);
        (_a = settings.userTestId) !== null && _a !== void 0 ? _a : (settings.userTestId = `${settings.testName}--${utils.general.guid()}`);
        settings.batch = { ...batch, ...settings.batch };
        (_b = (_1 = settings.batch).id) !== null && _b !== void 0 ? _b : (_1.id = (_c = utils.general.getEnvValue('BATCH_ID')) !== null && _c !== void 0 ? _c : `generated-${utils.general.guid()}`);
        (_d = (_2 = settings.batch).buildId) !== null && _d !== void 0 ? _d : (_2.buildId = utils.general.getEnvValue('BATCH_BUILD_ID'));
        (_e = (_3 = settings.batch).name) !== null && _e !== void 0 ? _e : (_3.name = utils.general.getEnvValue('BATCH_NAME'));
        (_f = (_4 = settings.batch).sequenceName) !== null && _f !== void 0 ? _f : (_4.sequenceName = utils.general.getEnvValue('BATCH_SEQUENCE'));
        (_g = (_5 = settings.batch).notifyOnCompletion) !== null && _g !== void 0 ? _g : (_5.notifyOnCompletion = utils.general.getEnvValue('BATCH_NOTIFY', 'boolean'));
        (_h = settings.keepBatchOpen) !== null && _h !== void 0 ? _h : (settings.keepBatchOpen = utils.general.getEnvValue('DONT_CLOSE_BATCHES', 'boolean'));
        (_j = settings.branchName) !== null && _j !== void 0 ? _j : (settings.branchName = (_k = utils.general.getEnvValue('BRANCH')) !== null && _k !== void 0 ? _k : utils.general.getEnvValue('BRANCH_NAME'));
        (_l = settings.parentBranchName) !== null && _l !== void 0 ? _l : (settings.parentBranchName = (_m = utils.general.getEnvValue('PARENT_BRANCH')) !== null && _m !== void 0 ? _m : utils.general.getEnvValue('PARENT_BRANCH_NAME'));
        (_o = settings.baselineBranchName) !== null && _o !== void 0 ? _o : (settings.baselineBranchName = (_p = utils.general.getEnvValue('BASELINE_BRANCH')) !== null && _p !== void 0 ? _p : utils.general.getEnvValue('BASELINE_BRANCH_NAME'));
        (_q = settings.gitBranchingTimestamp) !== null && _q !== void 0 ? _q : (settings.gitBranchingTimestamp = utils.general.getEnvValue('GIT_MERGE_BASE_TIMESTAMP'));
        (_r = settings.latestCommitInfo) !== null && _r !== void 0 ? _r : (settings.latestCommitInfo = getLatestCommitInfoFromEnvVars(logger));
        (_s = settings.ufgServerUrl) !== null && _s !== void 0 ? _s : (settings.ufgServerUrl = utils.general.getEnvValue('UFG_SERVER_URL'));
        (_t = settings.ignoreBaseline) !== null && _t !== void 0 ? _t : (settings.ignoreBaseline = false);
        (_u = settings.compareWithParentBranch) !== null && _u !== void 0 ? _u : (settings.compareWithParentBranch = false);
        (_v = settings.removeDuplicateTests) !== null && _v !== void 0 ? _v : (settings.removeDuplicateTests = removeDuplicateTests);
        (_w = settings.latestCommitInfo) !== null && _w !== void 0 ? _w : (settings.latestCommitInfo = await (0, extract_git_info_1.extractLatestCommitInfo)({ execOptions: { cwd }, logger }));
        if (settings.latestCommitInfo && !(0, extract_git_info_1.isISODate)(settings.latestCommitInfo.timestamp)) {
            logger.warn(`latestCommitInfo.timestamp is an invalid ISO date string: ${settings.latestCommitInfo.timestamp}`);
            settings.latestCommitInfo = undefined;
        }
        if (process.env.CI && (isRealBatchId(settings.batch.id) || settings.latestCommitInfo)) {
            await populateAutoScmInfo(isRealBatchId(settings.batch.id) ? settings.batch.id : settings.latestCommitInfo.sha);
        }
        if (settings.ignoreGitBranching) {
            settings.gitBranchingTimestamp = undefined;
        }
        else if (!settings.gitBranchingTimestamp) {
            const { branchName, parentBranchName } = settings;
            if (branchName && parentBranchName && branchName !== parentBranchName) {
                settings.gitBranchingTimestamp = await (0, extract_git_info_1.extractBranchingTimestamp)({
                    branchName,
                    parentBranchName,
                    execOptions: { cwd },
                    logger,
                });
            }
        }
        if (settings.gitBranchingTimestamp && !(0, extract_git_info_1.isISODate)(settings.gitBranchingTimestamp)) {
            logger.warn(`gitBranchingTimestamp is an invalid ISO date string: ${settings.gitBranchingTimestamp}`);
            settings.gitBranchingTimestamp = undefined;
        }
        const driver = target && (await (0, driver_1.makeDriver)({ spec, driver: target, logger, customConfig: settings }));
        const driverEnvironment = await (driver === null || driver === void 0 ? void 0 : driver.getEnvironment());
        const driverUrl = await (driver === null || driver === void 0 ? void 0 : driver.getDriverUrl());
        const multiDeviceTargets = [];
        for (const checkEnv of (_x = settings.environments) !== null && _x !== void 0 ? _x : []) {
            if (utils.types.has(checkEnv, 'iosDeviceInfo') && ((_y = driverEnvironment === null || driverEnvironment === void 0 ? void 0 : driverEnvironment.applitoolsLib) === null || _y === void 0 ? void 0 : _y.instrumented)) {
                multiDeviceTargets.push({
                    deviceInfo: checkEnv.iosDeviceInfo.deviceName,
                    orientation: (_z = checkEnv.iosDeviceInfo.screenOrientation) !== null && _z !== void 0 ? _z : 'portrait',
                });
            }
        }
        const account = await core.getAccountInfo({ settings: eyesServerSettings, logger });
        await handleBranchFallbackList(account, settings, logger);
        const componentConcurrency = account.serverConcurrency.componentConcurrency;
        settings.sessionConcurrency = handleConcurrency({
            userConcurrency: concurrency,
            componentConcurrency,
            useServerConcurrency,
            shouldPrintConcurrencyWarning,
            logger,
        });
        shouldPrintConcurrencyWarning = false;
        void core.logEvent({
            settings: {
                ...settings,
                level: 'Notice',
                event: {
                    type: 'openEyes',
                    userTestId: settings.userTestId,
                    concurrency: settings.sessionConcurrency,
                    environment,
                    driver: {
                        deviceName: driverEnvironment === null || driverEnvironment === void 0 ? void 0 : driverEnvironment.deviceName,
                        browserName: driverEnvironment === null || driverEnvironment === void 0 ? void 0 : driverEnvironment.browserName,
                        browserVersion: driverEnvironment === null || driverEnvironment === void 0 ? void 0 : driverEnvironment.browserVersion,
                        platformName: driverEnvironment === null || driverEnvironment === void 0 ? void 0 : driverEnvironment.platformName,
                        platformVersion: driverEnvironment === null || driverEnvironment === void 0 ? void 0 : driverEnvironment.platformVersion,
                        isApplitoolsLib: (_0 = driverEnvironment === null || driverEnvironment === void 0 ? void 0 : driverEnvironment.applitoolsLib) === null || _0 === void 0 ? void 0 : _0.instrumented,
                        isEC: driverEnvironment === null || driverEnvironment === void 0 ? void 0 : driverEnvironment.isEC,
                    },
                    driverUrl,
                    multiDeviceTargets,
                },
            },
            logger,
        });
        if (settings.disableBrokerUrlCache && driver) {
            await core.clearNMLClientCache(driver);
        }
        const getTypedEyes = (0, get_typed_eyes_1.makeGetTypedEyes)({
            type,
            settings: settings,
            target: driver,
            cores: cores !== null && cores !== void 0 ? cores : {
                ufg: (0, core_2.makeCore)({ spec, clients, base: core.base, asyncCache, logger }),
                classic: (0, core_1.makeCore)({ spec, clients, base: core.base, logger }),
            },
            logger,
        });
        const eyes = await getTypedEyes({ logger }).catch(error => {
            throw refineMissingApiKeyException(error, environment === null || environment === void 0 ? void 0 : environment.sdk);
        });
        return utils.general.extend(eyes, eyes => {
            var _a, _b;
            return ({
                getTypedEyes,
                check: (0, check_1.makeCheck)({ type, eyes, target: driver, spec, logger }),
                checkAndClose: (0, check_and_close_1.makeCheckAndClose)({ type, eyes, target: driver, spec, logger }),
                close: (0, close_1.makeClose)({ eyes, environments: (_a = config === null || config === void 0 ? void 0 : config.check) === null || _a === void 0 ? void 0 : _a.environments, logger }),
                abort: (0, abort_1.makeAbort)({ eyes, environments: (_b = config === null || config === void 0 ? void 0 : config.check) === null || _b === void 0 ? void 0 : _b.environments, logger }),
                getResults: (0, get_eyes_results_1.makeGetEyesResults)({ eyes, logger }),
            });
        }); // TODO solve the types issue
        /////// END FUNCTION BODY ///////////////////
        async function populateAutoScmInfo(batchId) {
            var _a, _b, _c, _d;
            var _e;
            const branchName = await (0, extract_git_info_1.extractGitBranch)({ execOptions: { cwd }, logger });
            const { owner: repoOwner, name: repoName } = await (0, extract_git_info_1.extractGitRepo)({ execOptions: { cwd }, logger });
            const buildId = await (0, extract_git_info_1.extractBuildIdFromCI)();
            if (branchName && repoOwner && repoName) {
                const scmInfo = await core.base
                    .updateIfScm({
                    settings: { ...eyesServerSettings, batchId, branchName, repoOwner, repoName, buildId },
                    logger,
                })
                    .catch(error => {
                    throw refineMissingApiKeyException(error, environment === null || environment === void 0 ? void 0 : environment.sdk);
                });
                if (scmInfo) {
                    (_a = settings.batch) !== null && _a !== void 0 ? _a : (settings.batch = {});
                    settings.batch.id = isRealBatchId(settings.batch.id) ? settings.batch.id : batchId;
                    (_b = (_e = settings.batch).buildId) !== null && _b !== void 0 ? _b : (_e.buildId = buildId);
                    // Takes data if it was not set by env. This is the priority according to yotam from BE
                    (_c = settings.branchName) !== null && _c !== void 0 ? _c : (settings.branchName = scmInfo.scmSourceBranch || undefined);
                    (_d = settings.parentBranchName) !== null && _d !== void 0 ? _d : (settings.parentBranchName = scmInfo.scmTargetBranch || undefined);
                }
            }
            else {
                logger.log(`Skipping check for SCM integration due to missing information: branchName=${branchName} repoName=${repoName} repoOwner=${repoOwner}`);
            }
        }
    };
    async function handleBranchFallbackList(account, settings, logger) {
        var _a, _b;
        const scm = getScmSettings(account.scmSettings);
        // Extract gitBranchName (the actual git branch name)
        if (!settings.ignoreGitBranching && scm.lookupFallbackList) {
            logger.log(`Extracting git branch name, ${cwd}`);
            settings.gitBranchName = await (0, extract_git_info_1.extractGitBranch)({
                execOptions: { cwd },
                logger,
                ignoreGitBranching: settings.ignoreGitBranching,
            });
            logger.log(`Extracted git branch name: ${settings.gitBranchName}`);
            (_a = settings.branchName) !== null && _a !== void 0 ? _a : (settings.branchName = settings.gitBranchName);
            // logger.log(`Using branch name: ${settings.branchName}`)
        }
        else {
            logger.log(`Skipping extraction of git branch name as ignoreGitBranching is set to true`);
        }
        if (settings.gitBranchName && !settings.ignoreGitBranching && scm.lookupFallbackList) {
            logger.log(`Extracting branch lookup fallback list for branch "${settings.gitBranchName}, ${cwd}"`);
            // add write to coralogix before and after
            settings.branchLookupFallbackList = await (0, extract_git_info_1.extractBranchLookupFallbackList)({
                gitBranchName: `${settings.gitBranchName}`,
                enableShallowClone: (_b = scm.enableShallowClone) !== null && _b !== void 0 ? _b : false,
                execOptions: { cwd },
                logger,
            });
        }
    }
}
exports.makeOpenEyes = makeOpenEyes;
function refineMissingApiKeyException(error, sdk) {
    var _a;
    if (!sdk || typeof sdk !== 'object')
        return error;
    try {
        if (error instanceof core_base_1.MissingApiKeyError) {
            const apiKeyTutorials = {
                python: {
                    'eyes-selenium': 'api-ref/sdk-api/selenium/python/#API',
                    'eyes-playwright': 'api-ref/sdk-api/playwright/python/#API',
                    'eyes-robotframework': 'api-ref/sdk-api/robot/#API',
                    'eyes-images': undefined,
                },
                ruby: {
                    eyes_images: 'api-ref/sdk-api/selenium/ruby/#API',
                    eyes_appium: undefined,
                    eyes_selenium: undefined,
                    eyes_universal: undefined,
                    eyes_capybara: undefined,
                },
                java: {
                    java_appium: undefined,
                    java_selenium: 'api-ref/sdk-api/selenium/java/#API',
                    java_playwright: 'api-ref/sdk-api/playwright/java/#API',
                    java_images: undefined,
                },
                dotnet: {
                    'eyes-selenium': 'api-ref/sdk-api/selenium/csharp/#API',
                    'eyes-playwright': 'api-ref/sdk-api/playwright/csharp/#API',
                    'eyes-appium': undefined,
                    'eyes-appium2': undefined,
                    'eyes-images': undefined,
                },
                js: {
                    'eyes-images': undefined,
                    'eyes-webdriverio': 'api-ref/sdk-api/wdio/#API',
                    'eyes-selenium': 'api-ref/sdk-api/selenium/javascript/#API',
                    'eyes-playwright': 'api-ref/sdk-api/playwright/javascript/#API',
                    'eyes-puppeteer': 'api-ref/sdk-api/puppeteer/javascript/#API',
                    'eyes-storybook': 'api-ref/sdk-api/storybook/#API',
                    'eyes-cypress': 'api-ref/sdk-api/cypress/#API',
                },
            };
            const { lang, name } = sdk !== null && sdk !== void 0 ? sdk : {};
            const path = (_a = apiKeyTutorials[lang]) === null || _a === void 0 ? void 0 : _a[name];
            if (path) {
                error.message = error.message.replace('Default.html#cshid=api', path);
            }
        }
    }
    catch {
        // if it fails, no harm done. The error message is fine without this refinement
        return error;
    }
    return error;
}
function getLatestCommitInfoFromEnvVars(logger) {
    const timestamp = utils.general.getEnvValue('GIT_LATEST_COMMIT_TIMESTAMP');
    const sha = utils.general.getEnvValue('GIT_LATEST_COMMIT_SHA');
    if (timestamp && sha) {
        return { timestamp, sha };
    }
    else if (timestamp && !sha) {
        logger.warn(`APPLITOOLS_GIT_LATEST_COMMIT_TIMESTAMP was provided without APPLITOOLS_GIT_LATEST_COMMIT_SHA. Proceeding provided timestamp ${timestamp} and NO_SHA_PROVIDED`);
        return { timestamp, sha: 'NO_SHA_PROVIDED' };
    }
    else if (sha && !timestamp) {
        logger.warn(`APPLITOOLS_GIT_LATEST_COMMIT_SHA was provided without APPLITOOLS_GIT_LATEST_COMMIT_TIMESTAMP. Not populating latestCommitInfo with the provided sha ${sha}.`);
    }
}
exports.getLatestCommitInfoFromEnvVars = getLatestCommitInfoFromEnvVars;
function isRealBatchId(batchId) {
    return !!batchId && !batchId.startsWith('generated');
}
function getScmSettings(scmSettings) {
    var _a, _b;
    // shallow clone is enabled by default and is currently not sent from the server
    let enableShallowClone = (_a = scmSettings.isShallowCloneEnabled) !== null && _a !== void 0 ? _a : true;
    let lookupFallbackList = (_b = scmSettings.isFallbackBranchListEnabled) !== null && _b !== void 0 ? _b : false;
    const envLookupFallbackList = utils.general.getEnvValue('ENABLE_BRANCH_LOOKUP_FALLBACK_LIST', 'boolean');
    const envShallowClone = utils.general.getEnvValue('SKIP_BRANCH_LOOKUP_IN_SHALLOW_CLONE', 'boolean');
    if (envLookupFallbackList !== undefined) {
        lookupFallbackList = envLookupFallbackList;
    }
    if (envShallowClone !== undefined) {
        enableShallowClone = !envShallowClone;
    }
    return {
        lookupFallbackList,
        enableShallowClone,
    };
}
function handleConcurrency({ userConcurrency, componentConcurrency, useServerConcurrency, shouldPrintConcurrencyWarning, logger, }) {
    let sessionConcurrency;
    // if user concurrency is 0 or negative, we ignore it and use server concurrency or default
    if (userConcurrency !== undefined && userConcurrency <= 0) {
        if (shouldPrintConcurrencyWarning) {
            logger.console.warn(`"testConcurrency" is invalid. Using server concurrency or default concurrency instead.`);
        }
        userConcurrency = undefined;
    }
    // TODO only eyes-storybook uses this hacky override `useServerConcurrency`, not all SDKs
    // remember when implementing for all SDKs: other SDKs use session concurrency, not component concurrency like storybook
    if (useServerConcurrency) {
        if (userConcurrency && userConcurrency < componentConcurrency) {
            if (shouldPrintConcurrencyWarning) {
                // TODO make this warning more generic
                logger.console.warn(`Eyes Storybook "testConcurrency" explicitly set to ${userConcurrency}, which is lower than your available limit of ${componentConcurrency}. Using the lower value as requested.`);
            }
            sessionConcurrency = userConcurrency;
        }
        else {
            sessionConcurrency = componentConcurrency;
            if (userConcurrency && shouldPrintConcurrencyWarning) {
                logger.console.warn(`Overriding user-defined "testConcurrency". Using organization concurrency limit from Eyes server: ${sessionConcurrency}`);
            }
        }
    }
    else {
        sessionConcurrency = userConcurrency !== null && userConcurrency !== void 0 ? userConcurrency : types_1.DEFAULT_CONCURRENCY;
    }
    return sessionConcurrency;
}
exports.handleConcurrency = handleConcurrency;
