// @applitools/req/dist/fetch-http2.js
import { fetch as undiciFetch, Agent } from 'undici';
import { Response as NodeFetchResponse } from 'node-fetch';
import { lookupWithCache } from './dns-cache.js';
import { Readable } from 'stream';
import { Buffer } from 'buffer';
import * as utils from '@applitools/utils';
const cachify = utils.general.cachify;
const cachifiedCreateAgent = cachify((options) => {
    return new Agent({
        allowH2: options.allowH2,
        maxHeaderSize: 16384,
        keepAliveTimeout: options.keepAliveTimeout,
        keepAliveMaxTimeout: options.keepAliveMaxTimeout,
        connect: { timeout: 60000, lookup: options.useDnsCache ? lookupWithCache : undefined },
        bodyTimeout: 300000,
        headersTimeout: 300000,
    });
});
const makeDispatcher = cachify(() => {
    return function dispatcher(url) {
        const isHttps = url.protocol === 'https:';
        // HTTP/2 requires TLS — only allow H2 for HTTPS URLs
        const allowH2 = isHttps;
        const useDnsCache = utils.general.getEnvValue('USE_DNS_CACHE') === 'true';
        const agentOptions = {
            allowH2,
            keepAliveTimeout: 30000,
            keepAliveMaxTimeout: 300000,
            useDnsCache,
        };
        return cachifiedCreateAgent(agentOptions);
    };
});
// --- OPTIMIZATION 1: Get Body Length without reading the body ---
function getBodyLength(url, body, headers) {
    if (url.size) {
        return String(url.size);
    }
    if (headers && (headers['content-length'] || headers['Content-Length'])) {
        return headers['content-length'] || headers['Content-Length'];
    }
    if (Buffer.isBuffer(body))
        return String(body.length);
    if (typeof body === 'string')
        return String(Buffer.byteLength(body));
    if (body && body.size)
        return String(body.size);
    return undefined;
}
export const fetchImpl = async (url, options = {}) => {
    var _a, _b, _c, _d, _e;
    let finalUrl;
    const finalOptions = { ...options };
    // 1. Parameter Normalization
    if (typeof url === 'object' && 'url' in url && !(url instanceof URL)) {
        const req = url; // Treated as NodeFetchRequest
        finalUrl = req.url;
        (_a = finalOptions.method) !== null && _a !== void 0 ? _a : (finalOptions.method = req.method);
        (_b = finalOptions.headers) !== null && _b !== void 0 ? _b : (finalOptions.headers = {});
        // Copy headers from the request object
        if (typeof ((_c = req.headers) === null || _c === void 0 ? void 0 : _c.forEach) === 'function') {
            req.headers.forEach((v, k) => {
                finalOptions.headers[k.toLowerCase()] = v;
            });
        }
        else {
            Object.assign(finalOptions.headers, req.headers);
        }
        // --- OPTIMIZATION 2: Don't Buffer. Use the stream directly. ---
        // If we haven't provided a body, use the one from the request.
        if (!finalOptions.body) {
            finalOptions.body = req.body;
        }
        (_d = finalOptions.timeout) !== null && _d !== void 0 ? _d : (finalOptions.timeout = req.timeout);
    }
    else {
        finalUrl = url;
    }
    // 2. Dispatcher Setup
    if (!finalOptions.dispatcher) {
        const urlObj = typeof finalUrl === 'string' ? new URL(finalUrl) : finalUrl;
        finalOptions.dispatcher = (_e = makeDispatcher()) === null || _e === void 0 ? void 0 : _e(urlObj);
    }
    // 3. Stream Handling (Crucial for Images)
    const isStream = finalOptions.body &&
        (typeof finalOptions.body.pipe === 'function' || typeof finalOptions.body[Symbol.asyncIterator] === 'function');
    if (isStream) {
        // Undici MUST have 'duplex: half' to stream a request body
        finalOptions.duplex = 'half';
    }
    const detectedSize = getBodyLength(url, finalOptions.body, finalOptions.headers);
    if (detectedSize) {
        // set content-length header only if it's not already set by the user, and if we were able to detect the size
        const contentLengthHeader = Object.keys(finalOptions.headers).find((h) => h.toLowerCase() === 'content-length');
        if (!contentLengthHeader) {
            finalOptions.headers = { ...finalOptions.headers, 'content-length': detectedSize };
        }
    }
    let timeoutId;
    let internalController;
    if (finalOptions.timeout) {
        internalController = new AbortController();
        timeoutId = setTimeout(() => internalController === null || internalController === void 0 ? void 0 : internalController.abort(), finalOptions.timeout);
        if (finalOptions.signal) {
            finalOptions.signal.addEventListener('abort', () => internalController === null || internalController === void 0 ? void 0 : internalController.abort(), { once: true });
        }
        finalOptions.signal = internalController.signal;
    }
    // Cleanup
    delete finalOptions.timeout;
    delete finalOptions.agent;
    delete finalOptions.size;
    try {
        const response = await undiciFetch(finalUrl, finalOptions);
        // Return stream to caller (Don't buffer response either!)
        const nodeBody = response.body ? Readable.from(response.body) : undefined;
        return new NodeFetchResponse(nodeBody, {
            status: response.status,
            statusText: response.statusText,
            headers: response.headers,
        });
    }
    finally {
        if (timeoutId)
            clearTimeout(timeoutId);
    }
};
// utility function to make transition from node-fetch to undici easier
// and to avoid converting streams to buffers just to get their length in fetch-http2 implementation
export function getBodySize(input, opts) {
    var _a, _b;
    // 1. Determine Method EARLY so we can decide if body is allowed
    const method = (_a = opts.method) !== null && _a !== void 0 ? _a : input.method;
    const isBodyAllowed = method && method.toUpperCase() !== 'GET' && method.toUpperCase() !== 'HEAD';
    let finalSize = undefined;
    let finalBody = (_b = opts.body) !== null && _b !== void 0 ? _b : input.body;
    if (isBodyAllowed) {
        if (opts.size) {
            finalSize = opts.size;
        }
        else {
            if (Buffer.isBuffer(finalBody)) {
                finalSize = finalBody.length;
            }
            else if (finalBody instanceof Uint8Array) {
                finalSize = finalBody.byteLength;
            }
            else if (typeof finalBody === 'string') {
                finalSize = Buffer.byteLength(finalBody);
            }
            else if (utils.types.isPlainObject(finalBody) || utils.types.isArray(finalBody) || finalBody === null) {
                finalBody = JSON.stringify(finalBody);
                finalSize = Buffer.byteLength(finalBody);
            }
        }
    }
    return finalSize;
}
