// @applitools/nml-client/dist/server/req-broker.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeReqBroker = void 0;
const req_1 = __importStar(require("@applitools/req"));
const utils = __importStar(require("@applitools/utils"));
function isInvalidBrokerMessage(result) {
    try {
        const parsedResult = JSON.parse(result);
        // Bare empty arrays indicate result is not ready yet
        if (Array.isArray(parsedResult) && parsedResult.length === 0)
            return true;
        // Check objects with empty payload arrays
        if (parsedResult && typeof parsedResult === 'object' && !Array.isArray(parsedResult)) {
            const { payload } = parsedResult;
            if (Array.isArray(payload) && payload.length === 0)
                return true;
        }
        return false;
    }
    catch {
        return true; // unparsable response is equivalent to empty one
    }
}
function makeReqBroker({ settings, logger }) {
    return (0, req_1.makeReq)({
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'User-Agent': settings.agentId,
        },
        proxy: settings.proxy,
        connectionTimeout: 300000 /* 5min */,
        retry: {
            limit: 5,
            timeout: process.env.EYES_NETWORK_RETRY_TIMEOUT ? Number(process.env.EYES_NETWORK_RETRY_TIMEOUT) : 200,
            codes: ['ECONNRESET', 'ECONNABORTED', 'ETIMEDOUT', 'ENOTFOUND', 'EAI_AGAIN', 'ENOMEM'],
        },
        hooks: [handleLogs({ logger }), handleLongRequests({ req: req_1.default, logger }), handleUnexpectedResponse()],
    });
}
exports.makeReqBroker = makeReqBroker;
function handleLogs({ logger: defaultLogger } = {}) {
    return {
        beforeRequest({ request, options }) {
            var _a;
            const logger = (_a = options === null || options === void 0 ? void 0 : options.logger) !== null && _a !== void 0 ? _a : defaultLogger;
            logger === null || logger === void 0 ? void 0 : logger.log(`Broker request "${options === null || options === void 0 ? void 0 : options.name}" will be sent to the address "[${request.method}]${request.url}" with body`, options === null || options === void 0 ? void 0 : options.body);
        },
        async afterResponse({ request, response, options }) {
            var _a;
            const logger = (_a = options === null || options === void 0 ? void 0 : options.logger) !== null && _a !== void 0 ? _a : defaultLogger;
            logger === null || logger === void 0 ? void 0 : logger.log(`Broker request "${options === null || options === void 0 ? void 0 : options.name}" that was sent to the address "[${request.method}]${request.url}" respond with ${response.statusText}(${response.status})`, response.status !== 200 ? `and body ${JSON.stringify(await response.clone().text())}` : '');
        },
        afterError({ request, error, options }) {
            var _a;
            const logger = (_a = options === null || options === void 0 ? void 0 : options.logger) !== null && _a !== void 0 ? _a : defaultLogger;
            logger === null || logger === void 0 ? void 0 : logger.error(`Broker request "${options === null || options === void 0 ? void 0 : options.name}" that was sent to the address "[${request.method}]${request.url}" failed with error`, error);
        },
    };
}
/**
 * Creates a custom fetch function that buffers polling responses to work around node-fetch clone bug.
 *
 * node-fetch 3.x has a bug where cloning a response after reading its body results in
 * empty clones. Since polling requests use retry validation that reads the body AND
 * have afterResponse hooks that also read the body, we need to buffer BEFORE retry
 * validation runs. We do this by wrapping the fetch function itself.
 *
 * See: js/packages/nml-client/docs/NODE_FETCH_CLONE_BUG.md for full investigation details
 */
function makeBufferingFetch() {
    return async (input, init) => {
        // Use globalReq to fetch, forwarding all options (proxy, headers, etc.)
        // Pass init directly to preserve proxy settings and other options
        const response = await (0, req_1.default)(input, init);
        // Buffer the body once
        const bufferedBody = await response.text();
        // Create fresh Response with buffered string (not stream)
        const bufferedResponse = new req_1.Response(bufferedBody, {
            status: response.status,
            statusText: response.statusText,
            headers: response.headers,
        });
        // Monkey-patch .text() and .json() to return from buffered string
        // WITHOUT consuming the Response body. This is critical because:
        // 1. Retry validation calls .text() to check for empty payloads
        // 2. Then hooks/caller call .json() to parse the result
        // 3. Calling .text() normally consumes the body, breaking .json()
        bufferedResponse.text = async () => bufferedBody;
        bufferedResponse.json = async () => JSON.parse(bufferedBody);
        // Also patch .clone() to return responses with the same monkey-patched methods
        const originalClone = bufferedResponse.clone.bind(bufferedResponse);
        bufferedResponse.clone = () => {
            const cloned = originalClone();
            cloned.text = async () => bufferedBody;
            cloned.json = async () => JSON.parse(bufferedBody);
            return cloned;
        };
        return bufferedResponse;
    };
}
function handleLongRequests({ req, logger: defaultLogger }) {
    return {
        async afterResponse({ request, response, options }) {
            var _a;
            if (response.status === 200) {
                let attemptCount = 0;
                const MAX_ATTEMPTS = 1500;
                const logger = (_a = options === null || options === void 0 ? void 0 : options.logger) !== null && _a !== void 0 ? _a : defaultLogger;
                // Use custom fetch that buffers responses to work around node-fetch clone bug
                const bufferingFetch = makeBufferingFetch();
                return req(request.url + '-response', {
                    proxy: options === null || options === void 0 ? void 0 : options.proxy,
                    fetch: bufferingFetch,
                    retry: [
                        {
                            // 1500 attempts x 200 ms = 5 minutes
                            limit: MAX_ATTEMPTS,
                            timeout: 200,
                            statuses: [404],
                        },
                        {
                            limit: MAX_ATTEMPTS,
                            timeout: 200,
                            validate: async ({ response }) => {
                                if (!response || response.status !== 200)
                                    return false;
                                try {
                                    const body = await response.clone().text();
                                    const isInvalid = isInvalidBrokerMessage(body);
                                    if (isInvalid) {
                                        logger === null || logger === void 0 ? void 0 : logger.log(`Broker polling request "${options === null || options === void 0 ? void 0 : options.name}" received empty response, retrying (attempt ${attemptCount + 1}/${MAX_ATTEMPTS})`);
                                    }
                                    return isInvalid;
                                }
                                catch (error) {
                                    return false;
                                }
                            },
                        },
                    ],
                    hooks: [
                        {
                            beforeRequest({ request, options: beforeOptions }) {
                                attemptCount++;
                                logger === null || logger === void 0 ? void 0 : logger.log(`Broker polling request "${options === null || options === void 0 ? void 0 : options.name}" attempt #${attemptCount}/${MAX_ATTEMPTS} will be sent to the address "[${request.method}]${request.url}" with body`, beforeOptions === null || beforeOptions === void 0 ? void 0 : beforeOptions.body);
                            },
                            async afterResponse({ request, response }) {
                                try {
                                    const body = await response.clone().text();
                                    logger === null || logger === void 0 ? void 0 : logger.log(`Broker polling request "${options === null || options === void 0 ? void 0 : options.name}" attempt #${attemptCount}/${MAX_ATTEMPTS} that was sent to the address "[${request.method}]${request.url}" respond with ${response.statusText}(${response.status}) with body ${JSON.stringify(body)}`);
                                }
                                catch (error) {
                                    logger === null || logger === void 0 ? void 0 : logger.log(`Broker polling request "${options === null || options === void 0 ? void 0 : options.name}" attempt #${attemptCount}/${MAX_ATTEMPTS} that was sent to the address "[${request.method}]${request.url}" respond with ${response.statusText}(${response.status})`);
                                }
                            },
                            afterError({ request, error }) {
                                var _a;
                                const logger = (_a = options === null || options === void 0 ? void 0 : options.logger) !== null && _a !== void 0 ? _a : defaultLogger;
                                logger === null || logger === void 0 ? void 0 : logger.error(`Broker polling request "${options === null || options === void 0 ? void 0 : options.name}" attempt #${attemptCount}/${MAX_ATTEMPTS} that was sent to the address "[${request.method}]${request.url}" failed with error`, error);
                            },
                        },
                    ],
                });
            }
        },
    };
}
function handleUnexpectedResponse() {
    return {
        async afterResponse({ response }) {
            if (response.status !== 200) {
                throw new Error(`Something went wrong when communicating with the mobile application, please try running your test again (error code: ${response.status})`);
            }
            const body = await response.clone().text();
            if (!body) {
                return response;
            }
            const result = JSON.parse(body);
            if (result === null || result === void 0 ? void 0 : result.payload) {
                const error = utils.types.isArray(result.payload)
                    ? result.payload.find(payload => payload.error)
                    : result.payload.error;
                if (error) {
                    const nmlError = new Error(`There was a problem when interacting with the mobile application. The provided error message was "${error.message}" and had a stack trace of "${error.stack}"`);
                    nmlError.nextPath = result.nextPath;
                    throw nmlError;
                }
            }
            return response;
        },
    };
}
