// @applitools/core/dist/autonomous/core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeAutonomousCore = void 0;
const logger_1 = require("@applitools/logger");
const core_1 = require("../ufg/core");
const take_snapshots_1 = require("./take-snapshots");
const create_render_target_1 = require("./create-render-target");
const start_renders_1 = require("./start-renders");
const get_render_results_1 = require("./get-render-results");
const create_render_results_1 = require("./create-render-results");
const open_eyes_1 = require("./open-eyes");
const open_check_and_close_1 = require("./open-check-and-close");
const register_default_scripts_1 = require("../ufg/utils/register-default-scripts");
const calculate_estimated_time_1 = require("./calculate-estimated-time");
function makeAutonomousCore({ spec, clients, base: defaultBase, asyncCache, fetchConcurrency, agentId = 'core-autonomous', logger: defaultLogger, } = {}) {
    const logger = (0, logger_1.makeLogger)({ logger: defaultLogger, format: { label: 'core-autonomous' } });
    const core = (0, core_1.makeCore)({ spec, clients, base: defaultBase, asyncCache, fetchConcurrency, agentId, logger });
    return {
        type: 'autonomous',
        takeSnapshots: spec && (0, take_snapshots_1.makeTakeSnapshots)({ core, spec, logger }),
        startSDKHook: spec && (0, register_default_scripts_1.makeRegisterOnNewDocumentScripts)({ spec, logger }),
        createRenderTarget: (0, create_render_target_1.makeCreateRenderTarget)({ core, logger }),
        startRenders: (0, start_renders_1.makeStartRenders)({ core, logger }),
        getRenderResults: (0, get_render_results_1.makeGetRenderResults)({ core, logger }),
        createRenderResults: (0, create_render_results_1.makeCreateRenderResults)({ logger }),
        /** @deprecated */
        openEyes: (0, open_eyes_1.makeOpenEyes)({ core, logger }),
        openCheckAndClose: (0, open_check_and_close_1.makeOpenCheckAndClose)({ core, logger }),
        calculateEstimatedTime: (0, calculate_estimated_time_1.makeCalculateEstimatedTime)({ core, logger }),
    };
}
exports.makeAutonomousCore = makeAutonomousCore;
