// @applitools/core-base/dist/server/req-eyes.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateOriginAndPath = exports.makeReqEyes = void 0;
const req_1 = __importStar(require("@applitools/req"));
const utils = __importStar(require("@applitools/utils"));
const path = __importStar(require("path"));
const missingApiKeyError_1 = require("../errors/missingApiKeyError");
const invalidApiKeyError_1 = require("../errors/invalidApiKeyError");
const mask_1 = require("../utils/mask");
const addRetryTimeout_1 = require("../utils/addRetryTimeout");
const backoffDelays = [
    1000, 1150, 1300, 1500, 1700, 2000, 2300, 2700, 3100, 3500, 4000, 4700, 5400, 6200, 7100, 8100, 9400, 10800, 12400,
    14200, 16400, 18800, 18800, 21600, 24900, 28600, 30000,
];
function makeReqEyes({ settings, fetch, logger }) {
    var _a, _b, _c, _d;
    const retryLimit = process.env.EYES_NETWORK_RETRY_LIMIT ? Number(process.env.EYES_NETWORK_RETRY_LIMIT) : 5;
    const retryTimeout = process.env.EYES_NETWORK_RETRY_TIMEOUT ? Number(process.env.EYES_NETWORK_RETRY_TIMEOUT) : 200;
    const retryDuration = process.env.EYES_NETWORK_RETRY_DURATION
        ? Number(process.env.EYES_NETWORK_RETRY_DURATION)
        : undefined;
    return (0, req_1.makeReq)({
        baseUrl: settings.eyesServerUrl,
        query: { removeSession: settings.removeSession },
        headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
            'x-applitools-eyes-client': settings.agentId,
            'User-Agent': settings.agentId,
            'X-Eyes-Api-Key': settings.apiKey,
        },
        proxy: settings.proxy,
        useDnsCache: settings.useDnsCache,
        connectionTimeout: (_a = settings.connectionTimeout) !== null && _a !== void 0 ? _a : 300000 /* 5min */,
        requestTimeout: (_b = settings.requestTimeout) !== null && _b !== void 0 ? _b : 30000,
        retryTimeout: (_c = settings.retryTimeout) !== null && _c !== void 0 ? _c : retryDuration,
        retry: (_d = settings.retry) !== null && _d !== void 0 ? _d : [
            // retry on network issues
            {
                limit: retryLimit,
                timeout: retryTimeout,
                codes: [
                    'ECONNRESET',
                    'ECONNABORTED',
                    'ECONNREFUSED',
                    'ETIMEDOUT',
                    'ENOTFOUND',
                    'EAI_AGAIN',
                    'STUCK_REQUEST',
                    'ENOMEM',
                    'UND_ERR_SOCKET',
                    'UND_ERR_CONNECT_TIMEOUT',
                    'UND_ERR_HEADERS_TIMEOUT',
                    'UND_ERR_BODY_TIMEOUT',
                ],
                validate(options) {
                    var _a, _b, _c, _d;
                    return ([404, 502, 503, 504, 500].includes((_b = (_a = options === null || options === void 0 ? void 0 : options.response) === null || _a === void 0 ? void 0 : _a.status) !== null && _b !== void 0 ? _b : 0) &&
                        !((_d = (_c = options === null || options === void 0 ? void 0 : options.response) === null || _c === void 0 ? void 0 : _c.headers) === null || _d === void 0 ? void 0 : _d.get('x-applitools-dont-retry')));
                },
            },
        ],
        hooks: [handleLongRequestsV2({ req: req_1.default, settings }), handleLogs({ logger }), handleUnexpectedResponse()],
        fetch,
    });
}
exports.makeReqEyes = makeReqEyes;
function handleLogs({ logger: defaultLogger } = {}) {
    const guid = utils.general.guid();
    let counter = 0;
    return {
        beforeRequest({ request, options }) {
            var _a;
            const logger = (_a = options === null || options === void 0 ? void 0 : options.logger) !== null && _a !== void 0 ? _a : defaultLogger;
            let requestId = request.headers.get('x-applitools-eyes-client-request-id');
            if (!requestId) {
                requestId = `${counter++}--${guid}`;
                request.headers.set('x-applitools-eyes-client-request-id', requestId);
            }
            logger === null || logger === void 0 ? void 0 : logger.log(`Request "${options === null || options === void 0 ? void 0 : options.name}" [${requestId}] will be sent to the address "[${request.method}]${request.url}" with body`, options === null || options === void 0 ? void 0 : options.body);
        },
        beforeRetry({ request, attempt, error, response, options }) {
            var _a;
            const logger = (_a = options === null || options === void 0 ? void 0 : options.logger) !== null && _a !== void 0 ? _a : defaultLogger;
            const requestId = request.headers.get('x-applitools-eyes-client-request-id');
            logger === null || logger === void 0 ? void 0 : logger.log(`Request "${options === null || options === void 0 ? void 0 : options.name}" [${requestId}] that was sent to the address "[${request.method}]${request.url}" with body`, options === null || options === void 0 ? void 0 : options.body, `is going to retried due to ${error ? 'an error' : 'a response with status'}`, error !== null && error !== void 0 ? error : `${response.statusText}(${response.status})`);
            request.headers.set('x-applitools-eyes-client-request-id', `${requestId.split('#', 1)[0]}#${attempt + 1}`);
        },
        async afterResponse({ request, response, options }) {
            var _a, _b, _c;
            const logger = (_a = options === null || options === void 0 ? void 0 : options.logger) !== null && _a !== void 0 ? _a : defaultLogger;
            const requestId = request.headers.get('x-applitools-eyes-client-request-id');
            const msg = `Request "${options === null || options === void 0 ? void 0 : options.name}" [${requestId}] that was sent to the address "[${request.method}]${request.url}" respond with ${response.statusText}(${response.status}), dont retry returned ${(_c = (_b = response === null || response === void 0 ? void 0 : response.headers) === null || _b === void 0 ? void 0 : _b.get('x-applitools-dont-retry')) !== null && _c !== void 0 ? _c : false}`;
            if (response.ok) {
                logger === null || logger === void 0 ? void 0 : logger.log(msg);
            }
            else {
                logger === null || logger === void 0 ? void 0 : logger.error(msg, `and body ${JSON.stringify(await response.clone().text())}`);
            }
        },
        afterError({ request, error, options }) {
            var _a;
            const logger = (_a = options === null || options === void 0 ? void 0 : options.logger) !== null && _a !== void 0 ? _a : defaultLogger;
            const requestId = request.headers.get('x-applitools-eyes-client-request-id');
            logger === null || logger === void 0 ? void 0 : logger.error(`Request "${options === null || options === void 0 ? void 0 : options.name}" [${requestId}] that was sent to the address "[${request.method}]${request.url}" failed with error`, error);
        },
    };
}
function handleUnexpectedResponse() {
    return {
        async afterResponse({ request, response, options }) {
            if ((options === null || options === void 0 ? void 0 : options.expected) &&
                (utils.types.isArray(options === null || options === void 0 ? void 0 : options.expected)
                    ? !options.expected.includes(response.status)
                    : options.expected !== response.status)) {
                const requestId = request.headers.get('x-applitools-eyes-client-request-id');
                const requestApiKey = request.headers.get('X-Eyes-Api-Key');
                const body = await response.clone().text();
                const bodyStr = body ? ` and body "${body}"` : '';
                if (response.status === 401) {
                    if (!requestApiKey || requestApiKey === 'undefined') {
                        throw new missingApiKeyError_1.MissingApiKeyError();
                    }
                    else {
                        throw new invalidApiKeyError_1.InvalidApiKeyError({ maskedApiKey: (0, mask_1.maskApiKey)(requestApiKey) });
                    }
                }
                else {
                    throw new Error(`Request "${options === null || options === void 0 ? void 0 : options.name}" [${requestId}] that was sent to the address "[${request.method}]${request.url}" failed due to unexpected status ${response.statusText}(${response.status})${bodyStr}`);
                }
            }
        },
    };
}
function handleLongRequestsV2({ req, settings }) {
    return {
        beforeRequest({ request }) {
            request.headers.set('Eyes-Expect-Version', '2');
            request.headers.set('Eyes-Expect', '202+result');
            request.headers.set('Eyes-Date', new Date().toUTCString());
        },
        async afterResponse({ response, options }) {
            var _a;
            const logger = options.logger;
            logger === null || logger === void 0 ? void 0 : logger.log(`Location header: ${response.headers.has('Location') ? response.headers.get('Location') : 'not found'}`);
            if (response.status === 202 && response.headers.has('Location')) {
                if (response.headers.has('Retry-After')) {
                    await utils.general.sleep(Number(response.headers.get('Retry-After')) * 1000);
                }
                const resultUrl = updateOriginAndPath(response.headers.get('Location'), settings.eyesServerUrl);
                // Use per-request retryTimeout if provided, otherwise fall back to settings.retryTimeout
                const currentRetryTimeout = (_a = options === null || options === void 0 ? void 0 : options.retryTimeout) !== null && _a !== void 0 ? _a : settings.retryTimeout;
                logger === null || logger === void 0 ? void 0 : logger.log(`Polling started for location: ${resultUrl}, with retryTimeout: ${currentRetryTimeout}`);
                const pollResponse = await req(resultUrl, options !== null && options !== void 0 ? options : {}, {
                    method: 'GET',
                    body: undefined,
                    ...(0, addRetryTimeout_1.addRetryTimeout)(currentRetryTimeout),
                    expected: [200, 201, 202, 203, 204, 205, 206, 207, 208, 502, 503, 504],
                    retry: {
                        timeout: backoffDelays,
                        validate: async ({ response }) => {
                            var _a;
                            return (!(response === null || response === void 0 ? void 0 : response.headers.get('x-applitools-dont-retry')) &&
                                ![400, 401, 403, 500].includes((_a = response === null || response === void 0 ? void 0 : response.status) !== null && _a !== void 0 ? _a : 400));
                        },
                    },
                    connectionTimeout: 0,
                    hooks: [
                        {
                            beforeRetry({ request, response, error: _error, attempt: _attempt }) {
                                if (response && [200, 201, 202].includes(response.status) && response.headers.has('Location')) {
                                    const location = updateOriginAndPath(response.headers.get('Location'), settings.eyesServerUrl);
                                    return { request, url: location };
                                }
                            },
                            afterError({ request, error, options }) {
                                const logger = options === null || options === void 0 ? void 0 : options.logger;
                                const requestId = request.headers.get('x-applitools-eyes-client-request-id');
                                logger === null || logger === void 0 ? void 0 : logger.error(`Request "${options === null || options === void 0 ? void 0 : options.name}" [${requestId}] that was sent to the address "[${request.method}]${request.url}" failed with error`, error);
                            },
                        },
                        handleUnexpectedResponse(),
                    ],
                });
                return pollResponse;
            }
        },
    };
}
function updateOriginAndPath(url, originalUrl) {
    const updatedUrl = new URL(url);
    const originalURL = new URL(originalUrl);
    if (updatedUrl.host === originalURL.host)
        return url;
    updatedUrl.protocol = originalURL.protocol;
    updatedUrl.host = originalURL.host;
    updatedUrl.pathname = path.posix.join(originalURL.pathname, updatedUrl.pathname);
    return updatedUrl.href;
}
exports.updateOriginAndPath = updateOriginAndPath;
