// @applitools/core-base/dist/heartbeat.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeHeartbeat = exports.clearAllHeartbeats = void 0;
const utils = __importStar(require("@applitools/utils"));
// Global state shared across all heartbeat instances
const heartbeats = new Map();
// Exported for testing purposes only - clears all heartbeat state
function clearAllHeartbeats() {
    for (const [apiKey, state] of heartbeats) {
        if (state.intervalId) {
            clearInterval(state.intervalId);
        }
        heartbeats.delete(apiKey);
    }
}
exports.clearAllHeartbeats = clearAllHeartbeats;
function makeHeartbeat({ requests, logger: mainLogger }) {
    const logger = mainLogger.extend(mainLogger, { tags: [`heartbeat-base-${utils.general.shortid()}`] });
    function stopHeartbeats(apiKey) {
        const state = heartbeats.get(apiKey);
        if (!state)
            return;
        logger.log('Stopping heartbeats for apiKey:', apiKey);
        if (state.intervalId) {
            clearInterval(state.intervalId);
        }
        heartbeats.delete(apiKey);
        logger.log('Heartbeat removed for apiKey:', apiKey);
    }
    async function startHeartbeats(apiKey, settings) {
        const state = heartbeats.get(apiKey);
        if (!state)
            return;
        // send the first heartbeat right away
        await requests.sendHeartbeat({ settings: { ...settings, processId: state.processId }, logger }).catch(error => {
            logger.error('Failed to send heartbeat:', error);
        });
        if (!heartbeats.has(apiKey)) {
            logger.log('Heartbeat state was removed during initial heartbeat for apiKey:', apiKey);
            return;
        }
        state.intervalId = setInterval(async () => {
            const currentState = heartbeats.get(apiKey);
            if (!currentState) {
                if (state.intervalId) {
                    clearInterval(state.intervalId);
                    logger.log('Cleared orphaned heartbeat interval for apiKey:', apiKey);
                }
                return;
            }
            try {
                await requests.sendHeartbeat({ settings: { ...settings, processId: currentState.processId }, logger });
            }
            catch (error) {
                if (error.message === 'This process is expired. All its sessions were abandoned') {
                    logger.log('Heartbeat process expired on server for apiKey:', apiKey);
                    // Process expired - stop and remove so next acquire will generate new processId
                    stopHeartbeats(apiKey);
                }
            }
        }, settings.interval);
        state.intervalId.unref();
    }
    function getOrCreateProcessId(apiKey) {
        let state = heartbeats.get(apiKey);
        if (state)
            return state.processId;
        const processId = utils.general.guid();
        state = {
            processId,
            intervalId: undefined,
            refCount: 0,
        };
        heartbeats.set(apiKey, state);
        logger.log('Created processId for apiKey:', apiKey, 'processId:', processId);
        return processId;
    }
    function acquire(settings) {
        const apiKey = settings.apiKey;
        const state = heartbeats.get(apiKey);
        if (!state) {
            throw new Error('Heartbeat state not found. getOrCreateProcessId must be called before acquire.');
        }
        const shouldStartHeartbeat = state.refCount === 0;
        state.refCount++;
        logger.log('Heartbeat acquired for apiKey:', apiKey, 'refCount:', state.refCount);
        if (shouldStartHeartbeat) {
            logger.log('Starting heartbeats for apiKey:', apiKey, 'with processId:', state.processId);
            void startHeartbeats(apiKey, settings).catch(error => {
                logger.error('Failed to start heartbeats:', error);
            });
        }
        return state.processId;
    }
    function release(apiKey) {
        const state = heartbeats.get(apiKey);
        if (!state) {
            logger.log('Heartbeat release called for unknown apiKey:', apiKey);
            return;
        }
        state.refCount--;
        logger.log('Heartbeat released for apiKey:', apiKey, 'refCount:', state.refCount);
        if (state.refCount === 0) {
            // Last session closed - stop heartbeat
            stopHeartbeats(apiKey);
        }
    }
    return {
        getOrCreateProcessId,
        acquire,
        release,
    };
}
exports.makeHeartbeat = makeHeartbeat;
