// @applitools/core-base/dist/open-eyes.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeOpenEyes = void 0;
const abort_controller_1 = require("abort-controller");
const check_1 = require("./check");
const check_and_close_1 = require("./check-and-close");
const close_1 = require("./close");
const abort_1 = require("./abort");
const get_results_1 = require("./get-results");
const utils = __importStar(require("@applitools/utils"));
const throttle_1 = require("./utils/throttle");
function makeOpenEyes({ requests, agentId: defaultAgentId, heartbeat, logger: mainLogger }) {
    let throttle;
    return async function openEyes({ settings, logger = mainLogger, }) {
        var _a;
        logger = logger.extend(mainLogger, { tags: [`eyes-base-${utils.general.shortid()}`] });
        if (!throttle) {
            throttle = (0, throttle_1.getThrottle)(settings.sessionConcurrency);
        }
        (_a = settings.processId) !== null && _a !== void 0 ? _a : (settings.processId = heartbeat.getOrCreateProcessId(settings.apiKey));
        settings.agentId = `${defaultAgentId} ${settings.agentId ? `[${settings.agentId}]` : ''}`.trim();
        logger.log('Command "openEyes" is called with settings', settings);
        return new Promise((resolve, reject) => {
            throttle(() => {
                return new Promise(async (done) => {
                    var _a;
                    try {
                        const controller = new abort_controller_1.AbortController();
                        const eyesRequests = await requests.openEyes({ settings, heartbeat, logger });
                        eyesRequests.test.ufgServer = {
                            ...eyesRequests.test.ufgServer,
                            ufgServerUrl: (_a = settings.ufgServerUrl) !== null && _a !== void 0 ? _a : eyesRequests.test.ufgServer.ufgServerUrl,
                        };
                        resolve(utils.general.extend(eyesRequests, {
                            check: (0, check_1.makeCheck)({ requests: eyesRequests, signal: controller.signal, logger }),
                            // eslint-disable-next-line prettier/prettier
                            checkAndClose: (0, check_and_close_1.makeCheckAndClose)({ requests: eyesRequests, done, signal: controller.signal, heartbeat, logger }),
                            close: (0, close_1.makeClose)({ requests: eyesRequests, done, heartbeat, logger }),
                            abort: (0, abort_1.makeAbort)({ requests: eyesRequests, done, controller, heartbeat, logger }),
                            getResults: (0, get_results_1.makeGetResults)({ requests: eyesRequests, logger }),
                        }));
                    }
                    catch (error) {
                        reject(error);
                        done();
                    }
                });
            });
        });
    };
}
exports.makeOpenEyes = makeOpenEyes;
