// @applitools/ufg-client/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserNameEnum = void 0;
var BrowserNameEnum;
(function (BrowserNameEnum) {
    BrowserNameEnum["CHROME"] = "chrome";
    BrowserNameEnum["CHROME_ONE_VERSION_BACK"] = "chrome-one-version-back";
    BrowserNameEnum["CHROME_TWO_VERSIONS_BACK"] = "chrome-two-versions-back";
    BrowserNameEnum["FIREFOX"] = "firefox";
    BrowserNameEnum["FIREFOX_ONE_VERSION_BACK"] = "firefox-one-version-back";
    BrowserNameEnum["FIREFOX_TWO_VERSIONS_BACK"] = "firefox-two-versions-back";
    BrowserNameEnum["IE_11"] = "ie";
    BrowserNameEnum["IE_10"] = "ie10";
    BrowserNameEnum["EDGE"] = "edge";
    BrowserNameEnum["EDGE_CHROMIUM"] = "edgechromium";
    BrowserNameEnum["EDGE_LEGACY"] = "edgelegacy";
    BrowserNameEnum["EDGE_CHROMIUM_ONE_VERSION_BACK"] = "edgechromium-one-version-back";
    BrowserNameEnum["EDGE_CHROMIUM_TWO_VERSIONS_BACK"] = "edgechromium-two-versions-back";
    BrowserNameEnum["SAFARI"] = "safari";
    BrowserNameEnum["SAFARI_ONE_VERSION_BACK"] = "safari-one-version-back";
    BrowserNameEnum["SAFARI_TWO_VERSIONS_BACK"] = "safari-two-versions-back";
})(BrowserNameEnum || (exports.BrowserNameEnum = BrowserNameEnum = {}));
